server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java

```diff
@@ -826,19 +827,34 @@ public TSFetchResultsResp fetchResults(TSFetchResultsReq req) {
       }
 
       QueryDataSet queryDataSet = queryId2DataSet.get(req.queryId);
-      TSQueryDataSet result = fillRpcReturnData(req.fetchSize, queryDataSet,
-          sessionIdUsernameMap.get(req.sessionId));
-
-      boolean hasResultSet = result.bufferForTime().limit() != 0;
-      if (!hasResultSet) {
-        queryId2DataSet.remove(req.queryId);
+      if (req.isAlign) {
+        TSQueryDataSet result = fillRpcReturnData(req.fetchSize, queryDataSet,
+            sessionIdUsernameMap.get(req.sessionId));
+        boolean hasResultSet = result.bufferForTime().limit() != 0;
+        if (!hasResultSet) {
+          queryId2DataSet.remove(req.queryId);
```

Comment:
Maybe you can call `QueryResourceManager.endQuery` here to release query resources ASAP.